dev/manual_tests/lib/focusable.dart

```diff
@@ -0,0 +1,208 @@
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+void main() {
+  runApp(MaterialApp(
+    title: 'Focusable Demo',
+    home: Scaffold(
+      appBar: AppBar(
+        title: const Text('Focusable Demo'),
+      ),
+      body: const Center(
+        child: FocusableDemo(),
+      ),
+    ),
+  ));
+}
+
+class FocusableButton extends StatefulWidget {
+  const FocusableButton({
+    Key key,
+    @required this.title,
+    this.autofocus = false,
+  }) : super(key: key);
+
+  final String title;
+  final bool autofocus;
+
+  @override
+  _FocusableButtonState createState() => _FocusableButtonState();
+}
+
+class _FocusableButtonState extends State<FocusableButton> {
+  bool _handleKeyEvent(FocusableNode node, RawKeyEvent event) {
+    if (event is RawKeyDownEvent) {
+      if (event.logicalKey == LogicalKeyboardKey.enter) {
+        _handleButtonPress();
+      }
+    }
+    print('Button ${widget.title} got key event $event');
+    return true;
+  }
+
+  void _handleButtonPress() {
+    print('Button ${widget.title} pressed.');
+  }
+
+  bool _hasFocus = false;
+
+  @override
+  Widget build(BuildContext context) {
+    print('Focusable Button ${widget.title} ${_hasFocus ? 'has focus' : 'does not have focus'}');
+    print(context.owner.focusManager.rootScope.toStringDeep());
+    return Focusable(
+      autofocus: widget.autofocus,
+      debugLabel: widget.title,
+      showDecorations: true,
+      onKey: _handleKeyEvent,
+      onFocusChange: (bool hasFocus) {
+        setState(() {
+          _hasFocus = hasFocus;
+        });
+      },
+      child: Builder(builder: (BuildContext context) {
+        return FlatButton(
+          onPressed: () {
+            Focusable.of(context).requestFocus();
+            _handleButtonPress();
+          },
+          child: Text(widget.title),
+        );
+      }),
+    );
+  }
+}
+
+class FocusableDemo extends StatefulWidget {
+  const FocusableDemo({Key key}) : super(key: key);
+
+  @override
+  _FocusableDemoState createState() => _FocusableDemoState();
+}
+
+class _FocusableDemoState extends State<FocusableDemo> {
+  @override
+  Widget build(BuildContext context) {
+    print('Built main widget: ${context.owner.focusManager.rootScope.toStringDeep()}');
+
+    final TextTheme textTheme = Theme.of(context).textTheme;
+    return Directionality(
+      textDirection: TextDirection.rtl,
+      child: DefaultFocusTraversal(
+        policy: const ReadingOrderTraversalPolicy(),
+        child: Focusable(
+          debugLabel: 'Scope',
+          isScope: true,
+          onKey: (FocusableNode node, RawKeyEvent event) {
+            if (event is RawKeyDownEvent) {
+              print('Scope got key event: ${event.logicalKey}, $node');
+              print('Keys down: ${RawKeyboard.instance.keysPressed}');
+              if (event.logicalKey == LogicalKeyboardKey.tab) {
```

Comment:
do you intend to put this logic into the UI libraries eventually? e.g. in the material and cupertino layers?

Reply:
Yes.

Reply:
I wanted to put out a breaking change announcement and get feedback before doing that (and to let us refine this some more before we switch over).